pipeline run command

droot opened the pull request:
This is a skeleton implemention for `kpt pipeline run` command. There are lot of places you will see `TODO` and notes and placeholder methods which may not be in the scope of the PR.

NOTE: I have test leaf level hydration manually on a pkg using `kpt pipeline run` command.

droot commented:
/cc @frankfarzan @mortent @Shell32-Natsu @natasha41575

droot commented:
/cc @mikebz

mortent commented on commands/pipelinecmd.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+// GetPipelineCommand returns a command that implements "kpt pipeline <run|...>" subcommands.
+func GetPipelineCommand(name string) *cobra.Command {
```
I think we should try to stick with the `runner`-pattern for the cobra commands: https://github.com/GoogleContainerTools/kpt/blob/655b215629ac693b736154a453dbf5a20d08e9a4/internal/cmdget/cmdget.go#L32

droot replied:
Ack. I started without Runner with a plan to add it when I need it, but didn't encounter any need. 
I think I will probably add it to wrap this command to include some auxiliary stuff such as logger, commandline options. Curious where else did you find runner useful ?

droot replied:
Moved the `run` subcommand to runner pattern as you suggested.

mortent commented on commands/pipelinecmd.go:
```diff
@@ -0,0 +1,118 @@
+// makeLogger returns a logger.
+// Primary reason for having this is to help in debugging.
+// TODO(droot): reconcile this with kpt logger (if one exists).
+func makeLogger() *zap.Logger {
```
What are the benefits of using the zap logging framework over klog?

droot replied:
We used zap in kubebuilder related projects and were sort of happy with it (it's very flexible and doesn't come in the way), so I used it here. Are we using any logging in kpt yet ? Haven't used klog. If klog solves all the problems of glog, then may be that is a better alternative.

mortent commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+// Executor hydrates a given pkg.
+type Executor struct {
+	PkgPath string
+	Logger  *zap.Logger
```
I'm thinking the logger should be passed through as a function parameter instead of being set on the structs. Maybe as a value on a context?

droot replied:
The downside of passing in context will be that every function will have to take a context, even the helper functions.
I will check klog and see if it can be a good alternative without requiring passing logger.

mortent commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+	if includeLocalResources {
+		var localResources []*yaml.RNode
+		// TODO(droot): figure out how to avoid reading sub packages here ?
+		pkgReader := &kio.LocalPackageReader{PackagePath: p.path, MatchFilesGlob: kio.MatchAll}
```
There is a property called `IncludeSubpackages` that I think should work.

droot replied:
Ack.

mortent commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+	Wet
+)
+
+func (s hydrationState) String() string {
```
Consider using the golang stringer here: https://godoc.org/golang.org/x/tools/cmd/stringer

droot replied:
hmmm...interesting. It does introduces dependency on `stringer` tool, wondering if it is worth the benefits (saving just one fn.)

mortent commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+	for _, s := range sources {
+		switch s {
+		case "./*":
+			// TODO(droot): figure out how to enumerate sub packages for current pkgPath.
```
I think this should work here: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/pathutil/pathutil.go#L14

droot replied:
Thanks.

frankfarzan replied:
That's looking for specific filename though.

You basically want `ls -d`. I foresee this getting more complicated in the future. For example, we may want to ignore 'hidden' directories (those start with `.`).

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+		}
+	}
+
+	if includeLocalResources {
```
We should extract out such common operations into functions or methods.

If going with OOP approach:

p.readLocalResources()

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+	// TODO(droot): parameterize the sink-mode (hctx.sinkMode) to determine whether to
+	// write it in-place or not. Currently in-place.
+	pkgWriter := &kio.LocalPackageWriter{PackagePath: p.path}
+	pipe := kio.Pipeline{
```
there's an overload of 'pipeline' here. Should probably trying to disambiguate through variable naming and comments.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+
+// expands takes a list of sources (./*, ./, ...) and current pkg path
+// and returns list of package paths that this pkg depends on.
+func expand(sources []string, pkgPath string) []string {
```
p.normalizedSources()?

Regardless of where validation logic runs, this needs to guarantee that source are normalized/clean/expanded. This logic will need to be well unit-tested and it will grow in complexity over time

frankfarzan replied:
sgtm

frankfarzan commented on internal/pipeline/fn.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+func (a *annotator) Run(r io.Reader, w io.Writer) error {
+	rw := &kio.ByteReadWriter{
```
Based on just this example, why isn't `rw' passed as an argument to this method?

frankfarzan replied:
If the function implementation always creates a rw from r and w and operates on that, that's a signal that we should just be passing rw as an argument?

frankfarzan commented on internal/pipeline/fn.go:
```diff
@@ -0,0 +1,182 @@
+// how to run itself. It accepts input from an io.Reader that is
+// a stream of KRM objects in the form of KRM FunctionSpec. It also accepts
+// a io.Rriter where it writes the result of running a KRM function.
+type KRMFn interface {
```
General question: Are you extracting a common library for use by 'kpt fn' and 'kpt pipeline' in the kpt repo?

droot reviewed (commented):
@frankfarzan Thanks for the review. Addressed majority of the comments.

Shell32-Natsu commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,290 @@
+		value: "test",
+	}
+
+	return []kio.Filter{
```
Maybe another TODO here for adding other functions from real pipeline.

droot replied:
Done.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+
+// Executor hydrates a given pkg.
+type Executor struct {
+	PkgPath string
```
https://github.com/GoogleContainerTools/kpt/pull/1340

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -0,0 +1,271 @@
+			includeLocalResources = true
+		} else {
+			// TODO(droot): recursive hydration is broken for now, needs more work
+			var depResources []*yaml.RNode
```
nit: "dep" and "dependency" are overloaded. Maybe 'transitiveResources'?

droot replied:
Done.

droot commented:
I think I have addressed most of the feedback. Treat this PR to be skeleton PR, so all the TODOs left in the PR will be done in the follow up PRs including the test coverage.

I think this lays out the shape of the solution very well and we can already see work done in other PRs fitting well in this skeleton. I would like to get this merged now so that @natasha41575 @Shell32-Natsu can start integrating their PRs and address TODOs in this PR.

^^ @frankfarzan @mortent

mortent commented on go.mod:
```diff
@@ -15,6 +15,7 @@ require (
 	github.com/spf13/cobra v1.0.0
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.6.1
+	go.uber.org/zap v1.10.0
```
I don't think this is still needed.

droot replied:
Yep, you are right. Just ran `go mod tidy` and it's gone.

mortent commented on internal/pipeline/cmd.go:
```diff
@@ -0,0 +1,66 @@
+		r.pkgPath = wd
+	} else {
+		// resolve and validate the provided path
+		r.pkgPath = args[0]
```
It seems like `args[0]` here will give us a relative path, while `os.Getwd()` will return an absolute path. I think we should always use absolute paths.

droot replied:
Down the line the pipeline.Executor is converting the path provided here to absolute path.

mortent reviewed (approved):
Some minor comments. We should consider adding tests early to make sure we have a design that lends itself to unit tests.

droot commented:
fixes #1321
